class-wc-calypso-bridge-dotcom-features.php

```diff
@@ -0,0 +1,110 @@
+<?php
+/**
+ * Functions for determining dotCom features in Calypso Bridge.
+ *
+ * @package WC_Calypso_Bridge/Classes
+ * @since   2.0.0
+ * @version 2.0.0
+ */
+
+defined( 'ABSPATH' ) || exit;
+
+if ( ! function_exists( 'wc_calypso_bridge_is_ecommerce_plan' ) ) {
+	/**
+	 * Returns if a site is an eCommerce plan site or not.
+	 *
+	 * @return bool True if the site is an ecommerce site.
+	 */
+	function wc_calypso_bridge_is_ecommerce_plan() {
+		return WC_Calypso_Bridge_DotCom_Features::is_ecommerce_plan();
+	}
+}
+
+if ( ! function_exists( 'wc_calypso_bridge_is_ecommerce_trial_plan' ) ) {
+	/**
+	 * Returns if a site is an eCommerce trial plan site or not.
+	 *
+	 * @return bool True if the site is an ecommerce trial site.
+	 */
+	function wc_calypso_bridge_is_ecommerce_trial_plan() {
+		return WC_Calypso_Bridge_DotCom_Features::is_ecommerce_trial_plan();
+	}
+}
+
+if ( ! function_exists( 'wc_calypso_bridge_is_business_plan' ) ) {
+
+	/**
+	 * Returns if a site is a Business plan site or not.
+	 *
+	 * @return bool True if the site is a business site.
+	 */
+	function wc_calypso_bridge_is_business_plan() {
+		return WC_Calypso_Bridge_DotCom_Features::is_business_plan();
+	}
+}
+
+/**
+ * WC Calypso Bridge DotCom Features class.
+ */
+class WC_Calypso_Bridge_DotCom_Features {
+
+	/**
+	 * Is Ecommerce plan.
+	 *
+	 * @var bool
+	 */
+	protected static $is_ecommerce_plan = null;
```

Comment:
Might it be cleaner to structure these variables and the related checks in a way that resembles the `WPCOM_Features` approach? We could even push the primary logic directly into `WPCOM_Features` so these are direct features in the system.

I ask this because `is_ecommerce_plan()` is already ambiguous given the existence of both the eCommerce and trial plans, and will only become more so as we introduce WooExpress plans. It won't be clear to future developers if `is_ecommerce_plan()` would reasonably include those plans or not.

Perhaps we can do this using naming that is more self-descriptive, like `$has_ecommerce_features` or `$has_ecommerce_optimizations`.
```suggestion
	protected static $has_ecommerce_features = null;
	protected static $has_ecommerce_optimizations = null;
```

Reply:
This seems tricky! 

> Might it be cleaner to structure these variables and the related checks in a way that resembles the WPCOM_Features approach? 

Could you elaborate on this? Are you thinking of a features-array that constructs different combinations? If so, we could revisit this class as we add plans+logic into the system. 

For now, I've added a static class helper to help us with the runtime caching and exposed the checks in global functions. I thought we could revisit the static class in the long term and always have the global functions as a way to interact with it. 

> I ask this because is_ecommerce_plan() is already ambiguous given the existence of both the eCommerce and trial plans, and will only become more so as we introduce WooExpress plans. It won't be clear to future developers if is_ecommerce_plan() would reasonably include those plans or not.

That's a great idea! The `wc_calypso_bridge_is_ecommerce_plan()` naming suggests this is a strict boolean check. I really liked the idea of renaming it to `wc_calypso_bridge_has_ecommerce_features()`. The only drawback is that the `wc_calypso_bridge_is_ecommerce_plan()` exists from the start. 

So moving to a different plan checking will require us to either 
- Introduce the new `wc_calypso_bridge_has_ecommerce_features()` and deprecate the previous one, or;
- Keep them both and make the `wc_calypso_bridge_is_ecommerce_plan()` strict. 

For example (pseudocode for demonstration):
```php
function wc_calypso_bridge_is_ecommerce_plan() {
    has_ecommerce_features && can_install_plugins;
}
```

Thoughts?

Reply:
I would also love to know more about the race condition where the `wpcom_site_has_feature()` won't be available! I think the best approach we can take is to do a sanity check before spinning up this entire project and bail early if we can't make any assumptions about the plan used. 

For example, we could add an early exit in the plugin's main file:
```php
// [...]
if ( ! function_exists( 'wpcom_site_has_feature' ) {
    return;
}

// The Bridge Main Controller.
require_once WC_CALYPSO_BRIDGE_PLUGIN_PATH . '/class-wc-calypso-bridge-dotcom-features.php';
require_once WC_CALYPSO_BRIDGE_PLUGIN_PATH . '/class-wc-calypso-bridge.php';
```